downstairs/src/region.rs

```diff
@@ -1270,6 +1346,75 @@ impl Extent {
 
         Ok(())
     }
+
+    /// Rehash the entire file. Remove any stored hash/encryption contexts
+    /// that do not correlate to data currently stored on disk. This is
+    /// primarily when opening an extent after recovering from a crash, since
+    /// irrelevant hashes will normally be cleared out during a flush().
+    ///
+    /// By default this function will only do work if the extent is marked
+    /// dirty. Set `force_override_dirty` to `true` to override this behavior.
+    /// This override should generally not be necessary, as the dirty flag
+    /// is set before any contexts are written.
+    #[instrument]
+    fn fully_rehash_and_clean_all_stale_contexts(
```

Comment:
No. Just as with the previous implementation of flush(), this only clears out contexts that do not match any existing on-disk block.

Reply:
Do we have a test for that case?  Maybe re-use the `multiple_context` test in `downstairs/src/region.rs` that
* Creates a region.
* Sets a few inner.contexts on some blocks, but nothing for others
* Create a full vec `extent_block_indexes_and_hashes` with hash data in every block
* Call `truncate_encryption_contexts_and_hashes` with the created vec above.
* Verify afterwards that the only contexts that exist are for the few we originally sent, that no other contexts exist.

I originally thought we would have to do this all in the integration suite area, but reading more I think we can
add a test around `truncate_encryption_contexts_and_hashes` that will verify this for us, and prevent
future us (most likely me) from doing something to break this required behavior.

Reply:
yeah so i actually modified `test_flush_removes_partial_writes` to test this instead. it can stand for a rename though. This makes sense to me because we're intentionally removing the behavior where flush removes partial writes, and moving it into the fully_rehash thing.

right now the changes i made have it close all the extents and re-open them to make sure it happens on extent open but we could flip that to an explicit call to `fully_rehash_...`

Reply:
Ah, yeah, `test_flush_removes_partial_writes()` is pretty good, but I think we can make it just
a little better, either in the same test or on another test. In addition to what there is currently, have
a test verify that: 
A totally unwritten block is still "unwritten", meaning a call to `region_write` with `only_write_unwritten` = `true`
will indeed write to the block.
A written block with multiple contexts is truncated to a single context, then we verify that a write_unwritten will not write to that block.  This would be the "normal" case where the truncate does what we want it to do.

Sorry for adding all this.   I'm very nervous about breaking something without knowing and having all these tests
help me sleep at night.